service/history/historyEngine.go

```diff
@@ -1593,6 +1582,208 @@ func (e *historyEngineImpl) SignalWorkflowExecution(signalRequest *h.SignalWorkf
+	}
+	release()
+
+Start_And_Signal: // Start workflow and signal
+	request := getStartRequest(sRequest)
+	if request.ExecutionStartToCloseTimeoutSeconds == nil || *request.ExecutionStartToCloseTimeoutSeconds <= 0 {
+		return nil, &workflow.BadRequestError{Message: "Missing or invalid ExecutionStartToCloseTimeoutSeconds."}
+	}
+	if request.TaskStartToCloseTimeoutSeconds == nil || *request.TaskStartToCloseTimeoutSeconds <= 0 {
+		return nil, &workflow.BadRequestError{Message: "Missing or invalid TaskStartToCloseTimeoutSeconds."}
+	}
+	execution = workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowId,
+		RunId:      common.StringPtr(uuid.New()),
+	}
+
+	// Generate first decision task event.
+	taskList := *request.TaskList.Name
+	msBuilder := newMutableStateBuilder(e.shard.GetConfig(), e.logger)
+	startedEvent := msBuilder.AddWorkflowExecutionStartedEvent(domainID, execution, request)
+	if startedEvent == nil {
+		return nil, &workflow.InternalServiceError{Message: "Failed to add workflow execution started event."}
+	}
+
+	if msBuilder.AddWorkflowExecutionSignaled(getSignalRequest(sRequest)) == nil {
+		return nil, &workflow.InternalServiceError{Message: "Failed to add workflow execution signaled event."}
+	}
+
+	var transferTasks []persistence.Task
+	decisionScheduleID := emptyEventID
+	decisionStartID := emptyEventID
+	decisionTimeout := int32(0)
+	di := msBuilder.AddDecisionTaskScheduledEvent()
+	if di == nil {
+		return nil, &workflow.InternalServiceError{Message: "Failed to add decision scheduled event."}
+	}
+
+	transferTasks = []persistence.Task{&persistence.DecisionTask{
+		DomainID: domainID, TaskList: taskList, ScheduleID: di.ScheduleID,
+	}}
+	decisionScheduleID = di.ScheduleID
+	decisionStartID = di.StartedID
+	decisionTimeout = di.DecisionTimeout
+
+	duration := time.Duration(*request.ExecutionStartToCloseTimeoutSeconds) * time.Second
+	timerTasks := []persistence.Task{&persistence.WorkflowTimeoutTask{
+		VisibilityTimestamp: e.shard.GetTimeSource().Now().Add(duration),
+	}}
+	// Serialize the history
+	serializedHistory, serializedError := msBuilder.hBuilder.Serialize()
+	if serializedError != nil {
+		logging.LogHistorySerializationErrorEvent(e.logger, serializedError, fmt.Sprintf(
+			"HistoryEventBatch serialization error on start workflow.  WorkflowID: %v, RunID: %v",
+			execution.GetWorkflowId(), execution.GetRunId()))
+		return nil, serializedError
+	}
+
+	err = e.shard.AppendHistoryEvents(&persistence.AppendHistoryEventsRequest{
+		DomainID:  domainID,
+		Execution: execution,
+		// It is ok to use 0 for TransactionID because RunID is unique so there are
+		// no potential duplicates to override.
+		TransactionID: 0,
+		FirstEventID:  startedEvent.GetEventId(),
+		Events:        serializedHistory,
+	})
+	if err != nil {
+		return nil, err
+	}
+	msBuilder.executionInfo.LastFirstEventID = startedEvent.GetEventId()
+
+	createWorkflow := func(isBrandNew bool, prevRunID string) (string, error) {
```

Comment:
this func does not take the workflow ID reuse policy in consideration.
moreover, it is hard to actually maintain a second set of start workflow logic.
so could we actually merge the signal & start to start workflow API?

Reply:
I think merging the implementation does make sense. But merging the APIs is not.

Reply:
i believe, history side start && start & signal APIs can be merged since it is not public.